### PR TITLE
request: separate request creation path from other objects

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Request objects are to performance critical and needs special handling. The PR let request object creation to have its own critical sections and also a different sets of indirect parameters.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test: mpich/warnings/quick

**Comment by hzhou** [comment]
test: mpich/warnings/quick

**Comment by hzhou** [comment]
test: mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/include/mpir_handlemem.h [hunk 1]
```diff
@@ -217,13 +217,14 @@ static inline void *MPIR_Handle_obj_alloc(MPIR_Object_alloc_t * objmem)
...
-    ret = MPIR_Handle_obj_alloc_unsafe(objmem);
+    ret = MPIR_Handle_obj_alloc_unsafe(objmem, HANDLE_NUM_BLOCKS, HANDLE_NUM_INDICES);
     MPID_THREAD_CS_EXIT(POBJ, MPIR_THREAD_POBJ_HANDLE_MUTEX);
     MPID_THREAD_CS_EXIT(VCI, MPIR_THREAD_VCI_HANDLE_MUTEX);
     return ret;
 }
 
-static inline void *MPIR_Handle_obj_alloc_unsafe(MPIR_Object_alloc_t * objmem)
+static inline void *MPIR_Handle_obj_alloc_unsafe(MPIR_Object_alloc_t * objmem,
+                                                 int max_blocks, int max_indices)
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_handlemem.h:227]: Glad to.

File: src/include/mpir_objects.h [hunk 2]
```diff
@@ -421,6 +421,12 @@ typedef struct MPIR_Object_alloc_t {
     void *direct;               /* Pointer to direct block, used
                                  * for allocation */
     int direct_size;            /* Size of direct block */
+    void *lock;                 /* lower-layer may regiter a lock to use. This is
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_objects.h:424]: thx.

File: src/include/mpir_objects.h [hunk 3]
```diff
@@ -421,6 +421,12 @@ typedef struct MPIR_Object_alloc_t {
...
+                                 * mostly for multipool requests. For other objects
+                                 * or not per-vci thread granularity, this lock
+                                 * pointer is ignored. Ref. mpir_request.h.
+                                 * NOTE: it is `void *` because mutex type not defined yet.
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_objects.h:428]: I'll need you clarify your question (maybe by making it more specific or with example).

File: src/include/mpir_request.h [hunk 4]
```diff
@@ -346,6 +347,11 @@ static inline MPIR_Request *MPIR_Request_create(MPIR_Request_kind_t kind, int po
     return req;
 }
 
+static inline MPIR_Request *MPIR_Request_create(MPIR_Request_kind_t kind)
+{
+    return MPIR_Request_create_pool(kind, 0);
+}
+
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h:373]: I lost track which patch this is. I will keep your note to make sure it works with `git bisect`

File: src/include/mpir_objects.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (wesbland) [reply src/include/mpir_objects.h:428]: In the comment, you say this needs to be void because the real data type isn’t defined yet. Can it be defined before this?

File: src/include/mpir_request.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (wesbland) [reply src/include/mpir_request.h:373]: It’s the revert patch and the one right before it.

File: src/include/mpir_objects.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_objects.h:428]: The inclusion order is defined in `mpiimpl.h`. The mutex are wrapped by `MPID_` so it can't be pulled ahead in a "device-independent" header.

[the remark this answers is not in the document]
Comment (wesbland) [reply src/include/mpir_objects.h:428]: I see. That's fine then.

File: src/include/mpir_handlemem.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_handlemem.h:235]: Comment added.

File: src/include/mpir_request.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h:373]: I think this is another github ordering mess. I believe it is fixed after I re-date the commits.

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/async/ticketlock

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/async/ticketlock

File: src/include/mpir_request.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/include/mpir_request.h:375]: From what I can tell, this patch won't compile. It changes the request creation API without updating any callers.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h:375]: Should I squash with the "Revert..." commit? Or I can make sure to squash it before merge. I was trying to make it easier to review.

[the remark this answers is not in the document]
Comment (raffenet) [reply src/include/mpir_request.h:375]: Yeah, I think squash with a note that the patch includes a revert of [abc123].

File: src/include/mpir_request.h [hunk 5]
```diff
@@ -424,8 +440,10 @@ static inline void MPIR_Request_free(MPIR_Request * req)
 
         int pool = (req->handle & REQUEST_POOL_MASK) >> REQUEST_POOL_SHIFT;
         MPID_THREAD_CS_ENTER(POBJ, MPIR_THREAD_POBJ_HANDLE_MUTEX);
+        MPID_THREAD_CS_ENTER(VCI, (*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock));
```
Comment (raffenet) [inline src/include/mpir_request.h]: This VCI critical section is removed in `request: remove the VCI handle mutex for request` and then added back here 2 commits later. Is this intentional?

File: src/include/mpir_request.h [hunk 6]
```diff
@@ -265,7 +265,7 @@ static inline void MPII_init_request(void)
  * with each pool that they are going to use, typically a 1-1 vci-pool mapping.
  * NOTE: currently, only per-vci thread granularity utilizes multiple request pool.
  */
-static inline void MPIR_Request_registor_pool_lock(int pool, MPID_Thread_mutex_t * lock)
+static inline void MPIR_Request_register_pool_lock(int pool, MPID_Thread_mutex_t * lock)
```
Comment (raffenet) [inline src/include/mpir_request.h:268]: nit: unrelated typo fix

File: src/include/mpir_request.h (hunk 5 above)
Comment (hzhou) [reply src/include/mpir_request.h]: `VCI handle mutex` is global for all object handle mem, here we need per-pool request mutex.

File: src/include/mpir_request.h (hunk 6 above)
Comment (hzhou) [reply src/include/mpir_request.h:268]: It must be commited to the wrong patch. I'll check

File: src/include/mpir_request.h [hunk 7]
```diff
@@ -392,7 +392,7 @@ MPL_STATIC_INLINE_PREFIX MPIR_Request *MPIR_Request_create_complete(MPIR_Request
     return req;
 }
 
-static inline void MPIR_Request_free(MPIR_Request * req)
+static inline void MPIR_Request_free_with_safety(MPIR_Request * req, int need_safety)
```
Comment (raffenet) [inline src/include/mpir_request.h:397]: I think I'd rather see us use the safe/unsafe convention vs. this API. Not sure what others think.

File: src/include/mpir_request.h [hunk 8]
```diff
@@ -301,6 +301,8 @@ static inline MPIR_Request *MPIR_Request_create_pool(MPIR_Request_kind_t kind, i
 {
     MPIR_Request *req;
 
+    /* hack for testing */
+    MPIDU_THREAD_ASSERT_IN_CS((*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock));
```
Comment (raffenet) [inline src/include/mpir_request.h:305]: This would be good to leave in for `--enable-g` builds, probably.

Comment (hzhou) [reply src/include/mpir_request.h:305]: Good idea. There are quite a few thread-related assertions that are really meant for debugging. Maybe we should add an option `--enable-g=thread`?

File: src/include/mpir_request.h (hunk 7 above)
Comment (hzhou) [reply src/include/mpir_request.h:397]: Only in the `ch4` inside the device layer we have the case of device using the same lock for both operations and request pool, therefore, needs explicit interface to differentiate the cases. Outside `ch4`, there is really no case for `unsafe` version. Do you really want to expose this complications to all code paths that doesn't apply?

File: src/include/mpir_request.h (hunk 8 above)
Comment (raffenet) [reply src/include/mpir_request.h:305]: we have these options already, though I do not know their state of functionality
```
        mutex    - Enable error checking on pthread mutexes
        mutexnesting - Check for non-nesting of mutexes
```

File: src/include/mpir_request.h (hunk 7 above)
Comment (raffenet) [reply src/include/mpir_request.h:397]: Hmm, would it make sense to make this an internal `ch4` API then? I don't really want to churn ch3 callers, no.

File: src/include/mpir_request.h (hunk 8 above)
Comment (hzhou) [reply src/include/mpir_request.h:305]: Never aware of these 😄 I think we can combine the, both into `mutex`?

File: src/include/mpir_request.h (hunk 7 above)
Comment (hzhou) [reply src/include/mpir_request.h:397]: The whole point is not to churn `ch3` and `mpir`. They just use `MPIR_Request_free` as before.

File: src/include/mpir_request.h (hunk 8 above)
Comment (raffenet) [reply src/include/mpir_request.h:305]: I think so. `mutexnesting` doesn't appear to have any affect, anyway.

File: src/include/mpir_request.h (hunk 7 above)
Comment (raffenet) [reply src/include/mpir_request.h:397]: Right. I guess what I am saying is leave `MPIR_Request_free` as it is and add an `MPIDI_Request_free_unsafe` in ch4 only. Not sure if that is more hacky or not. Would be good to get some additional eyeballs on this commit, IMO.

Comment (hzhou) [reply src/include/mpir_request.h:397]: The thing is, not all calls from ch4 is `unsafe` call. My current view is to keep code outside `ch4` non-vci-aware, but inside ch4, vci-aware. So use `MPIR_Request_free_with_safety` is my intention to keep the code inside explicitly aware.

Comment (hzhou) [reply src/include/mpir_request.h:397]: We can't call within `ch4` like following either:
```
MPID_THREAD_CS_ENTER
MPIR_Request_free_unsafe()
MPID_THREAD_CS_EXIT
```
Because many of `MPIR_Request_free` inside ch4 doesn't really free the request, so don't need incur the critical section.

File: src/include/mpir_request.h (hunk 5 above)
Comment (raffenet) [reply src/include/mpir_request.h]: OK, I see it now.

File: src/include/mpir_request.h (hunk 7 above)
Comment (raffenet) [reply src/include/mpir_request.h:397]: So just going by `git grep MPIR_Request_free src/mpid/ch4`, all usages of `MPIR_Request_free_with_safety` are unsafe. There are 2 remaining usages of `MPIR_Request_free` (safe). One in `MPID_Request_free_hook` and the other in `MPIDI_workq_release_pt2pt_elemt` (untested?). Am I missing anything?

I'm not suggesting changing the `MPIR_Request_free` API to `MPIR_Request_free_safe`. Rather, leave it as-is for legacy and add an "unsafe" API for use where applicable. Or we *could* rename `MPIR_Request_free`->`MPIR_Request_free_safe` and then just macro replace it in `ch3` for compatibility.

Comment (hzhou) [reply src/include/mpir_request.h:397]: > There are 2 remaining usages of `MPIR_Request_free` (safe). One in `MPID_Request_free_hook` and the other in `MPIDI_workq_release_pt2pt_elemt` (untested?). Am I missing anything?

Not right now. PR #4530 is going to move the critical section from ch4-layer down to netmod/shmmod/am layer, then all ch4-layer usage will need the safe version.

Note that the safety is added deep inside `MPIR_Request_free_with_safety` when it actually need releasing the handle memory, so the safety has to be a parameter. I guess I can create another wrapper function `MPIR_Request_free_unsafe` which is `MPIR_Request_free_with_safety(req, 0)`. And maybe we can create one more wrapper `MPIR_Request_free_safe` either as macro or inline wrapper for ch4-usage. Would you prefer that?

Comment (raffenet) [reply src/include/mpir_request.h:397]: I initially made my comment because the "with_safety" API went against safe/unsafe convention already used in MPICH. Your point about the actual locking being within a conditional does make it harder to just simply define safe and unsafe versions. That is too bad 😦.

Comment (hzhou) [reply src/include/mpir_request.h:397]: That was the reason I need `with_safety` routines, but now think about it, I have no problem make that internal routines, create `free/free_safe/free_unsafe` wrappers, all wrap around the internal `with_safety` call.

File: src/include/mpir_request.h (hunk 6 above)
Comment (hzhou) [reply src/include/mpir_request.h:268]: fixed.

File: src/include/mpir_request.h (hunk 7 above)
Comment (hzhou) [reply src/include/mpir_request.h:398]: Added and used `MPIR_Reqeust_free_safe/MPIR_Request_free_unsafe` wrappers.

File: src/include/mpir_request.h (hunk 8 above)
Comment (hzhou) [reply src/include/mpir_request.h:305]: Left the assertion inside `#ifdef MPICH_DEBUG_MUTEX`-guard. Also followed convention and renamed the macro `MPID_THREAD_ASSERT_IN_CS(VCI, mutex)`.

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=mutex
netmod: ch4:ofi

File: src/include/mpir_request.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h:376]: Squashed

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=mutex
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=mutex
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=mutex
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=mutex
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/async/ticketlock

**Comment by hzhou** [comment]
test:mpich/async/ticketlock

File: src/mpid/ch4/src/ch4_recv.h [hunk 9]
```diff
@@ -373,7 +373,7 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_RECV_INIT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_RECV_INIT);
 
-    rreq = MPIR_Request_create_from_pool(MPIR_REQUEST_KIND__PREQUEST_RECV, 0);
+    rreq = MPIR_Request_create(MPIR_REQUEST_KIND__PREQUEST_RECV);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_recv.h:376]: I'm curious why the difference here vs. in the workq commit [6fc64009541a47a04ed0463d3f4041b918fd0e4b] which adds a VCI lock around `MPIR_Request_create`. There seems to be a contradiction with this statement:

> Now that we don't automatically add critical section inside
    MPIR_Request_create

Comment (hzhou) [reply src/mpid/ch4/src/ch4_recv.h:376]: Ahh... good question! I did that early on to get `workq` out-of-my-way before the later iterations. We should either drop this commit or change the `MPIR_Request_create` into `MPIR_Request_create_pool`. I think the latter is probably better since the workq is inside ch4 and it will need use vci explicitly at some point.

Comment (hzhou) [reply src/mpid/ch4/src/ch4_recv.h:376]: But your question was about the `prequest`, that is because prequest never enters device, so it doesn't need care about `vci`.

Comment (hzhou) [reply src/mpid/ch4/src/ch4_recv.h:376]: > Ahh... good question! I did that early on to get `workq` out-of-my-way before the later iterations. We should either drop this commit or change the `MPIR_Request_create` into `MPIR_Request_create_pool`. I think the latter is probably better since the workq is inside ch4 and it will need use vci explicitly at some point.

Actually I did use `MPIR_Request_create_from_pool` for the workq in https://github.com/pmodels/mpich/pull/4499/commits/f837c3cdcb3c36f8ac46da120ee68b333a039927, so it is good :)

Comment (hzhou) [reply src/mpid/ch4/src/ch4_recv.h:376]: There was some reasoning to differentiate prequest and workq -- mainly one is mpi-layer concept with less flexibility, and the other can be fully tweaked. However, in principle they have the same issues. Let me drop that commit and use explicit critical section in stead. That way, at least we have consistency inside `ch4`.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpid/ch4/src/ch4_recv.h (hunk 9 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_recv.h]: @raffenet I dropped this commit and made all these request creation inside ch4 consistent, i.e. with explicit critical section and `MPIR_Request_create_from_pool`.

**Comment by hzhou** [comment]
> We might want to squash some more of these commits together to avoid runtime errors introduced by the changes. For instance the use of pool-based request allocation in each ch4 component.

The four commits that replace ch4-layer `MPIR_Request_free_unsafe` are squashed with the api change https://github.com/pmodels/mpich/pull/4499/commits/fdeb139c99d2ce07d567f6a450f226dcb626e42a

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
This PR broke master (with `--enable-g`).  It needs to be reverted till it can be fixed.

**Comment by hzhou** [comment]
> This PR broke master (with `--enable-g`). It needs to be reverted till it can be fixed.

Do you have a jenkins link?

**Comment by pavanbalaji** [comment]
No, jenkins reviews don't test `--enable-g`, which is probably how it got in in the first place.

**Comment by hzhou** [comment]
> No, jenkins reviews don't test `--enable-g`, which is probably how it got in in the first place.

The "config" build is with `--enable-g=all`, right? https://pmrs-jenkins.cels.anl.gov/view/mpich-review/job/mpich-review-ch4-ofi/1272/

Do you mean with `-enable-g` without `=` part?

**Comment by pavanbalaji** [comment]
> > No, jenkins reviews don't test `--enable-g`, which is probably how it got in in the first place.
> 
> The "debug" build is with `--enable-g=all`, right? https://pmrs-jenkins.cels.anl.gov/view/mpich-review/job/mpich-review-ch4-ofi/1272/
> 
> Do you mean with `-enable-g` without `=` part?

The failure I saw was with `ch3`.  There was `VCI` used in the MPIR layer.  Here's the only snippet of the error that I have remaining (sorry I overwrote the build):

```
In file included from ../mpich/src/include/mpiimpl.h:225:0,
                 from ../mpich/src/mpi/attr/attr_delete.c:6:
../mpich/src/include/mpir_request.h: In function 'MPIR_Request_create_from_pool':
../mpich/src/include/mpir_request.h:305:5: warning: implicit declaration of function 'MPID_THREAD_ASSERT_IN_CS' [-Wimplicit-function-declaration]
     MPID_THREAD_ASSERT_IN_CS(VCI, (*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock));
     ^
../mpich/src/include/mpir_request.h:305:30: error: 'VCI' undeclared (first use in this function)
     MPID_THREAD_ASSERT_IN_CS(VCI, (*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock));
                              ^
../mpich/src/include/mpir_request.h:305:30: note: each undeclared identifier is reported only once for each function it appears in
../mpich/src/include/mpir_request.h: In function 'MPIR_Request_free_with_safety':
../mpich/src/include/mpir_request.h:453:38: error: 'VCI' undeclared (first use in this function)
             MPID_THREAD_ASSERT_IN_CS(VCI, (*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock));
                                      ^
make[2]: *** [src/mpi/attr/lib_libmpi_la-attr_delete.lo] Error 1
make[2]: Leaving directory `/sandbox/balaji/mpich/build'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/sandbox/balaji/mpich/build'
make: *** [all] Error 2
```

**Comment by hzhou** [comment]
Thanks, that's sufficient.

**Comment by hzhou** [comment]
Pushed PR #4572 to fix it
